packages/admin/src/error/errordialog/ErrorDialogProvider.tsx

```diff
@@ -0,0 +1,21 @@
+import * as React from "react";
+
+import ErrorDialog, { ErrorDialogComponentRefType, ErrorDialogOptions } from "./ErrorDialog";
+import { ErrorDialogContext } from "./ErrorDialogContext";
+
+export const ErrorDialogProvider: React.FunctionComponent = ({ children }) => {
+    const errorDialogRef = React.useRef<ErrorDialogComponentRefType>(null);
```

Comment:
why not lift up the state into this component instead of this relatively complicated api?

Reply:
yeah, looks like a complicated api, but in my opinion it's necessary. 

The problem is, the circularity of updating the states and rerendering of `ApolloProvider` and `ErrorDialogProvider`.

We had this kind of problems by another project. Everything looked really nice and as soon as an apollo error occurred, react continually started to rerender.

The Problem is that the JSX was like:

```
<ErrorDialogProvider>
     <ApolloProvider>
          <Application which causes an apollo error/>
    </ApolloProvider
</ErrorDialogProvider>
```

the `ApolloProvider` uses the Context of the `ErrorDialogContext`, so each time the state of the `ErrorDialogContext` changes the ApolloProvider gets rerendered, and the query gets requeried and the error occurred again, the state of the `ErrorDialogContext` gets changed again and we got a rerendering loop


we looked for a possible solution and we fixes this by moving the state from the `ErrorDialogContext` to a child Component `ErrorDialog`. The resulting JSX looks like:

```
<ErrorDialogProvider>
     <ApolloProvider>
          <Application which causes an apollo error/>
    </ApolloProvider
    <ErrorDialog />
</ErrorDialogProvider>
```

with this jsx structure, when the state gets handled in the `ErrorDialog` Component and changes, it only rerenders the children of `ErrorDialog` and it does not affect the other children (in this case the ApolloProvider).


Do you know another, simple or better solution to this problem?

Reply:
I think I found a good way:
```
const ChildrenContainer = React.memo(({ children }) => {
  return <>{children}</>;
});


<ErrorDialogProvider>
     <ChildrenContainer>{children}</ChildrenContainer>
    <ErrorDialog />
</ErrorDialogProvider>
```
Demo: https://codesandbox.io/s/goofy-wescoff-ypq1r

Reply:
the demo looks good, but i think the example is too simple because there is one more player, and that's the ApolloProvider. I will try this out, and maybe i provide another example

Reply:
just tried it out again. Unfortunately it's like i said and the suggested solution does not work. 

**See more:**
Original Broken Solution: https://github.com/vivid-planet/comet-admin/tree/error-dialog-endless-loading-endless-error
Suggested Solution: https://github.com/vivid-planet/comet-admin/tree/error-dialog-endless-loading-suggested-solution 

![Screen Recording 2021-05-18 at 20 43 10 mov](https://user-images.githubusercontent.com/6098356/118706572-cee8f980-b819-11eb-9ca2-9cd1a5891bb9.gif)

i try to explain the problem once again:

This are the playersPlayer:
- ErrorDialogContext
- ApolloProvider
- useErrorDialog
- Component which gets has Broken Query or Server error,...


**RenderPasses:**
1. Render ErrorDialogContext
2. Render ApolloProvider
3. Render Componente with Query
4. Query in Component produces Error
5. ApolloProvider -> onErrorLink gets triggerd (be aware, ApolloProvider uses useErrorDialog Hook -> and will get rerendered if the context values changes)
6. on ErrorLink says ErrorDialogContext to show Error -> which sets state in ErrorDialogContext 
7. as soon as the state gets set in the ErrorDialogContext, everything gets rerendered and starts again on 2. - **loop**

Reply:
fixed it :)
https://github.com/vivid-planet/comet-admin/commit/0a29c7d7a49724dcb4303c8c315933021bf4ca7d
(I pushed into error-dialog-endless-loading-suggested-solution branch)

Reply:
the rerender problem is solved - cool 👍 , but now the alert does not show up.....

Reply:
i will take a look in it

Reply:
@nsams i am sorry, your solution does not work either. 

You wrapped the showError Method to often, so the passed options did't got set in the state because they always where null. So the state didn't trigger a rerender at all, and the rerender problem didn't occur.


Simplified Story and add logs, and fixed showErrorMethod https://github.com/vivid-planet/comet-admin/commit/2d94d88cdf772d248259a7476b971cd7d99ad159#diff-1622679b3611d0ddb74405b573870364c5b070115a8e73801f7c834957f20278L19

Reply:
oops, I used useCallback completely wrong - the technique is the right one, tough.

problem was, that memoization of the function alone doesn't help at all, as a new errorDialogContext value still gets re-created on every render. The whole thing needs to be memoized: 97fb6dae33d8cac27d1a80b2882540b3b95017fc

Reply:
@nsams looks https://github.com/vivid-planet/comet-admin/commit/97fb6dae33d8cac27d1a80b2882540b3b95017fc looks good for me. Thank you for the help!

i will try to get this into the MR

Reply:
works like a charm 😍